lib/stdlib/src/gen_statem.erl

```diff
@@ -845,109 +850,117 @@ loop_hibernate(Parent, Debug, S) ->
        {wakeup_from_hibernate,3}}).
 
 %% Entry point for wakeup_from_hibernate/3
-loop_receive(Parent, Debug, S) ->
+loop_receive(Parent, Debug, #{cancel_timers := 0, auto_hibernate_timeout := AutoHibernateTimeout} = S) when is_integer(AutoHibernateTimeout) ->
     receive
 	Msg ->
-	    case Msg of
-		{system,Pid,Req} ->
-		    #{hibernate := Hibernate} = S,
-		    %% Does not return but tail recursively calls
-		    %% system_continue/3 that jumps to loop/3
-		    sys:handle_system_msg(
-		      Req, Pid, Parent, ?MODULE, Debug, S,
-		      Hibernate);
-		{'EXIT',Parent,Reason} = EXIT ->
-		    %% EXIT is not a 2-tuple therefore
-		    %% not an event but this will stand out
-		    %% in the crash report...
-		    Q = [EXIT],
-		    terminate(exit, Reason, ?STACKTRACE(), Debug, S, Q);
-		{timeout,TimerRef,TimerMsg} ->
-		    #{timer_refs := TimerRefs,
-		      timer_types := TimerTypes,
-		      hibernate := Hibernate} = S,
-		    case TimerRefs of
-			#{TimerRef := TimerType} ->
-			    %% We know of this timer; is it a running
-			    %% timer or a timer being cancelled that
-			    %% managed to send a late timeout message?
-			    case TimerTypes of
+		handle_received_msg(Msg, Parent, Debug, S)
+	after
+	    AutoHibernateTimeout ->
+		    loop_hibernate(Parent, Debug, S)
+    end;
+loop_receive(Parent, Debug, S) ->
+	receive
+		Msg ->
+			handle_received_msg(Msg, Parent, Debug, S)
+	end.
+
+handle_received_msg({system,Pid,Req}, Parent, Debug, S) ->
```

Comment:
If you still want to keep this function maybe because you want to match on messages in function clauses instead of a `case` clause for style reasons; it should be renamed since there is a comment higher up that claims that all functions in the server loop are named `loop_*` (does not apply to helper functions).

I suggest e.g `loop_receive_msg/4`, and the argument order should be `(Parent, Debug, S, Msg)` since tail call (jump) with the same argument order as the function you jump from does not need to shuffle around the registers and is therefore faster.  So add and remove arguments at the end of the argument list when jumping a loop.